File: schemas/cms-blog/convex/menu_items.ts
// menu_items: Hierarchical navigation links within a menu.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const menuItems = defineTable({
  menuId: v.id("menus"),
  parentId: v.optional(v.id("menu_items")),
  label: v.string(),
  linkType: v.union(
    v.literal("post"),
    v.literal("category"),
    v.literal("custom")
  ),
  linkPostId: v.optional(v.id("posts")),
  linkCategoryId: v.optional(v.id("categories")),
  linkUrl: v.optional(v.string()),
  openInNewTab: v.boolean(),
  sortOrder: v.number(),
  isActive: v.boolean(),
  updatedAt: v.number(),
})
  .index("by_menu_id_parent_id_sort_order", [
    "menuId",
    "parentId",
    "sortOrder",
  ])
  .index("by_link_post_id", ["linkPostId"])
  .index("by_link_category_id", ["linkCategoryId"]);

File: schemas/cms-blog/spacetimedb/menu_items.rs
// menu_items: Individual navigation links within a menu hierarchy.
// See README.md for full design rationale.
use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum MenuLinkType {
    Post,     // type: String
    Category,
    Custom,
}

#[spacetimedb::table(name = menu_items, public)]
pub struct MenuItem {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub menu_id: String, // FK -> menus.id (cascade delete)
    pub parent_id: Option<String>, // FK -> menu_items.id (cascade delete)
    pub label: String,
    pub link_type: MenuLinkType,
    #[index(btree)]
    pub link_post_id: Option<String>, // FK -> posts.id (cascade delete)
    #[index(btree)]
    pub link_category_id: Option<String>, // FK -> categories.id (cascade delete)
    pub link_url: Option<String>,
    pub open_in_new_tab: bool,
    pub sort_order: i32, // Composite index: (menu_id, parent_id, sort_order)
    pub is_active: bool,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
